src/main/python/main/ayab/engine/websocketserial.py

```
@@ -0,0 +1,66 @@
+                    break
+
+            data = self._rxbuffer[0:size]
+            self._rxbuffer = self._rxbuffer[size:]
+        return data
+
+    def write(self, data: bytes) -> None:
+        try:
+            with self._lock:
+                self._ws.send(data)
+        except websockets.ConnectionClosed:
+            raise ConnectionError("WebSocket connection is closed")        
```

Ruff (0.11.9)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)